crates/matrix-sdk/src/widget_api/messages.rs

```diff
@@ -0,0 +1,90 @@
+use serde::{Deserialize, Serialize};
+
+use super::{capabilities::Options, handler::ApiVersion};
+
+// #[derive(Serialize, Deserialize, Debug)]
+// #[serde(tag = "api")]
+// pub enum JsonMessage {
+//     #[serde(rename = "fromWidget")]
+//     FromWidget(FromWidget),
+//     #[serde(rename = "toWidget")]
+//     ToWidget(ToWidget),
+// }
+
+#[derive(Debug, Serialize, Clone)]
+#[serde(tag = "action")]
+pub enum ToWidget {
+    // Requests.
+    #[serde(rename = "capabilities")]
+    SendMeCapabilities,
+    #[serde(rename = "notify_capabilities")]
+    CapabilitiesUpdated,
+
+    // Responses.
+    #[serde(rename = "supported_api_versions")]
+    SupportedApiVersions(Message<Vec<ApiVersion>, ()>),
+    #[serde(rename = "content_loaded")]
+    ContentLoaded(Message<(), ()>),
+    #[serde(rename = "org.matrix.msc2931.navigate")]
+    Navigate(Message<String, ()>),
+}
+
+#[derive(Debug, Deserialize, Clone)]
+#[serde(tag = "action")]
+pub enum FromWidget {
+    // Requests.
+    #[serde(rename = "supported_api_versions")]
+    SupportedApiVersions(Message<(), ()>),
+    #[serde(rename = "content_loaded")]
+    ContentLoaded(Message<(), ()>),
+    #[serde(rename = "org.matrix.msc2931.navigate")]
+    Navigate(Message<String, ()>),
+
+    // Responses.
+    #[serde(rename = "capabilities")]
+    Capabilities(Message<(), Options>),
+    #[serde(rename = "notify_capabilities")]
+    CapabilitiesConfirmed(Message<(), ()>),
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Message<Req, Resp> {
+    #[serde(flatten)]
+    pub header: Header,
+    #[serde(rename = "data")]
+    pub request: Req,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub response: Option<Response<Resp>>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Header {
+    pub request_id: String,
+    pub widget_id: String,
+}
```

Comment:
I think it indeed does not give any advantages right now. I think it was just my OCD to split things up so that I can see which things could be re-used (I was not exactly sure if I'll end up having 2 structures or less).

However, it feels like `Header` is a separate part logically in a sense that I can imagine that there will be a `Transport` that generates the `Header` since we don't really deal with such things as `request_id` in the high level logic, so it's likely that there will be a component that deals with these sort of things.